ports/qemu-riscv/Makefile

```diff
@@ -0,0 +1,140 @@
+include ../../py/mkenv.mk
+-include mpconfigport.mk
+
+# qstr definitions (must come before including py.mk)
+QSTR_DEFS = qstrdefsport.h
+
+# MicroPython feature configurations
+MICROPY_ROM_TEXT_COMPRESSION ?= 1
+
+# include py core make definitions
+include $(TOP)/py/py.mk
+include $(TOP)/extmod/extmod.mk
+
+BOARD ?= virt
+
+CROSS_COMPILE ?= riscv64-unknown-elf-
+
+# The GCC version string starts like this:
+# <target>-gcc (<optional git hash>) <major>.<minor>.<patch>
+GCC_VERSION = $(subst ., , $(word 3, $(shell $(CC) --version)))
+GCC_MAJOR = $(word 1, $(GCC_VERSION))
+GCC_MINOR = $(word 2, $(GCC_VERSION))
+GCC_PATCH = $(word 3, $(GCC_VERSION))
+
+ifeq ($(BOARD),virt)
+ABI=ilp32
+# GCC 10 and lower do not recognise the Zicsr extension in the
+# architecture name.  "Make" unfortunately does not provide any simple way
+# to perform numeric comparisons, so to keep things simple we assume that
+# GCC is at least version 10 for the time being.
+ifeq ($(GCC_MAJOR),10)
+ARCH=rv32imac
+else
+# Recent GCC versions explicitly require to declare extensions.
+ARCH=rv32imac_zicsr
+endif
+AFLAGS = -mabi=$(ABI) -march=$(ARCH)
+CFLAGS += -mabi=$(ABI) -march=$(ARCH) -mcmodel=medany
+CFLAGS += -DQEMU_SOC_VIRT
+ifeq ($(PICOLIBC),1)
+CFLAGS += -DUSE_PICOLIBC -D__TMP_FLT_EVAL_METHOD
```

Comment:
Picolibc will define [`__PICOLIBC__`](https://github.com/picolibc/picolibc/blob/695583c3b7dc461204948db5c8390e967b526a27/picolibc.h.in#L465) so can test for this instead of `USE_PICOLIBC` where needed.

If the float eval method macro needs to be defined then maybe that can happen in mphalport.h or similar, gated behind `__PICOLIBC__`)? Do you know the details of why this needs to be passed in?

If that's all possible, then the last special-case for picolibc is having to pass the libc header path in from the command line CFLAGS. It feels like there's probably some multilib magic that can resolve this, at which point the same build command will work for both...?

Reply:
It's been a while, but `__TMP_FLT_EVAL_METHOD` was needed due to a symbol conflict between `picolibc` and `libm`, so that had to be defined in the makefile rather than in the source files themselves:

```bash
In file included from ../../lib/libm/math.c:27:
../../lib/libm/libm.h:22: error: "FLT_EVAL_METHOD" redefined [-Werror]
   22 | #define FLT_EVAL_METHOD 0
      | 
In file included from /usr/lib/picolibc/riscv64-unknown-elf/include/sys/config.h:34,
                 from /usr/lib/picolibc/riscv64-unknown-elf/include/_ansi.h:44,
                 from /usr/lib/picolibc/riscv64-unknown-elf/include/ieeefp.h:37,
                 from /usr/lib/picolibc/riscv64-unknown-elf/include/math.h:39,
                 from ../../lib/libm/libm.h:20,
                 from ../../lib/libm/math.c:27:
/usr/lib/gcc/riscv64-unknown-elf/10.2.0/include/float.h:197: note: this is the location of the previous definition
  197 | #define FLT_EVAL_METHOD __FLT_EVAL_METHOD__
      | 
cc1: all warnings being treated as errors
```

That said, a simpler yet nastier workaround could be to always define `__TMP_FLT_EVAL_METHOD` and then rely on `__PICOLIBC__` as you suggest.  Could something like that still be acceptable?  (I'm still wondering why Ubuntu didn't ship newlib for RISC-V like it did for other targets...)

Reply:
Thanks for the explanation, that's really useful. If this is happening in MicroPython's libm header then do you think we can work around it in there with some kind of `#ifndef __PICOLIBC__` guard? This way it doesn't bleed out to the top-level Makefile (it seems like no other ports use picolibc at this time, so shouldn't cause any regression.) I still don't fully understand what this selects, though...

> (I'm still wondering why Ubuntu didn't ship newlib for RISC-V like it did for other targets...)

It is indeed frustrating when stuff like this breaks, but honestly IMHO picolibc is long overdue. newlib-nano has a bunch of architectural decisions that don't really suit small micros, and picolibc improves a lot of stuff while retaining libc compatibility. Keith Packard has done a pretty impressive job with it, tackling decades of technical debt for libc on embedded.

So I guess I grudgingly respect Ubuntu for "pulling the bandaid off", here. :). It does add an extra dimension for testing now though, as there's now two possible libcs under the hood.

Reply:
Well, if picolibc is here to stay (and given keithp's track record it will), then such a define could do.  However I'd rather have it outside this PR to not step outside the PR's scope, so if you don't have the time to file it I can do that :)

That will then get rid of those two extra defines, and since I found a workaround for having the build work out of the box on Ubuntu 22.04 as well it would sort all issues out I guess!

Reply:
Just one small addendum - since we don't include `picolibc.h` directly, we cannot use `__PICOLIBC__` for feature detection (and we can't define it ourselves, or the build will fail due to a duplicated definition).

How about a new `mpconfigport.h` flag (`MICROPY_PICOLIBC`) instead?

Reply:
> since we don't include picolibc.h directly, we cannot use __PICOLIBC__ for feature detection (and we can't define it ourselves, or the build will fail due to a duplicated definition).

You should be able to include any libc header, this will include sys/features.h and that will include picolibc.h on picolibc at which point the macro is defined! :magic_wand: 

If that seems too magic, you can include sys/features.h directly - this header is not standard C but it's a newlib thing, so it'll be there on both supported libcs.

Reply:
Sorry, I got that wrong, I was checking for `__PICOLIBC__` but it should have been `_PICOLIBC__` (one leading underscore instead of two).  Things work indeed as you explained.

Reply:
By the way, I've looked a bit into `FLT_EVAL_METHOD`.  In theory it should only be set by the compiler, as indicates whether floats and doubles are evaluated as such or both assumed as floats or both as doubles.

As a test, I removed the `FLT_EVAL_METHOD` define in `lib/libm/libm.h` and it built just fine, with all tests passing.  I've tried the same for the `qemu-arm` port (that uses newlib) and it built with GCC 10 and passed all tests as well.

I see it's been added in f04329e93b7bfa5462f232832d64aac3dc52d5db to fix tests on the pyboard.  I assume code was built using C90 as its C standard version, as `FLT_EVAL_METHOD` was added in C99.

So, maybe the fix would be to remove the entry rather than `#ifndef` it.  If backwards compatiblity with code built in C90 mode is needed, then another option would be to move the `#define` before the inclusion of `math.h` as both picolibc and newlib define that symbol if it wasn't defined by the compiler itself first.

Reply:
Thanks very much for digging into that, it's very informative! I don't think supporting pre-C99 is a concern, but we should probably put up a separate PR to remove this and check for any unexpected consequences.

Reply:
> Sorry, I got that wrong, I was checking for __PICOLIBC__ but it should have been _PICOLIBC__ (one leading underscore instead of two). Things work indeed as you explained.

Oh, I got that wrong too in that case. Anyway, glad you've got it working.

Reply:
> So maybe we should static assert on sizeof(float) somewhere in the implementations, just to be sure that's always the case. I can put up a PR for this if you like.

That'd be great, thank you.  In the meantime I'll file a PR for `libm`/`libm_dbl` changes.